### PR TITLE
Nearest neighbors for training set

### PR DESCRIPTION
**Pull request by dataplayer12**
Hi, this is in response to the issue I raised.

### CODE REVIEW

**Comment by erikbern** [comment]
Nice. Can you fix two things?

1. You just renamed the datasets in HDF5. ann-benchmarks uses those names in other places so you'd have to rename them elsewhere too. Or just keep the name of the current datasets and name so that it's `distances` and `train_distances` (a bit inconsistent maybe, but seems fine)
2. There's a bit of code redundancy between generating the train neighbors and test neighbors. Can you put it in a for loop or use a function or something to reduce the code duplication?

**Comment by erikbern** [comment]
I also wonder how much this will inflate the size of the indexes. I'll take a look. Maybe not a big issue.

**Comment by dataplayer12** [comment]
I have added a function to avoid redundancy. I think it would be okay if we skip storing distances for train data. Since the indices of 100 neighbors are provided, the time to compute distances of NNs is still reduced by a factor of 10^4.

**Comment by erikbern** [comment]
Looks good. Was thinking actually what could make sense is to store the top 100 neighbors and distances for all vectors in the test set, but only 10 for the ones in the train set, just so it takes a bit less memory. What do you think?

**Comment by dataplayer12** [comment]
10 points is too few to calculate commonly used metrics like recall and mAP. If you are concerned about memory, I can enable chunking in the h5 file so that only a small chunk of the data is read at any time. Currently you store all 1M points in one huge contiguous array. Let me know.

**Comment by erikbern** [comment]
Let's do 100 then, but let's skip the distances.

**Comment by dataplayer12** [comment]
Thanks. In that case, the code should work as is. Look forward to the updated datasets.

**Comment by erikbern** [comment]
Will try to run this at some point. I'm just a bit concerned it will be incredibly slow. It might take days/weeks to run. I'll kick it off and see if it's doable.
